R/teal_slice.R

```diff
@@ -268,22 +268,60 @@ c.teal_slice <- function(...) {
   ans
 }
 
+#' @keywords internal
+#' @export
+as.list.teal_slice <- function(x) {
+  checkmate::assert_class(x, "teal_slice")
+
+  formals <- if (inherits(x, "teal_slice_expr")) {
+    formals(filter_expr)
+  } else {
+    formals(filter_var)
+  }
+
+  x <- if (shiny::isRunning()) {
+    shiny::reactiveValuesToList(x)
+  } else {
+    shiny::isolate(shiny::reactiveValuesToList(x))
+  }
+
+  formal_args <- setdiff(names(formals), '...')
+  extra_args <- setdiff(names(x), formal_args)
+
+  x[c(formal_args, extra_args)]
+}
+
 
-# format method for teal_slice
 #' @export
+#' @param show_all `logical(1)` should parameters set to NULL be returned
+#' @param center `logical(1)` should the output be centered and trimmed
 #' @rdname teal_slice
 #' @keywords internal
 #'
-format.teal_slice <- function(x, show_all = FALSE, ...) {
+format.teal_slice <- function(x, show_all = FALSE, center = TRUE, ...) {
   checkmate::assert_flag(show_all)
+  checkmate::assert_flag(center)
 
-  x <- if (shiny::isRunning()) {
-    rev(shiny::reactiveValuesToList(x))
-  } else {
-    rev(shiny::isolate(shiny::reactiveValuesToList(x)))
-  }
+  x_list <- as.list(x)
+  if (!show_all) x_list <- Filter(Negate(is.null), x_list)
+
+  x_json <- jsonlite::toJSON(x_list, pretty = TRUE, auto_unbox = TRUE, digits = 16, null = "null")
+  x_json_s <- strsplit(x_json, split = '\n')[[1]]
+
+  if (center) x_json_s <- center_json(x_json_s)
+
+  paste(c("teal_slice", x_json_s), collapse = "\n")
+}
+
+# centering of json output for `teal_slices` object JSON representation
+#' @param json a result of `jsonlite::toJSON(as.list(teal_slice), pretty = TRUE, auto_unbox = TRUE)`
+#' @keywords internal
+#'
+center_json <- function(json) {
```

Comment:
I suggest `center_json` be rewritten something like this:
```
center_json <- function(json) {
  format_name <- function(n) {
    if (n %in% c("{", "}")) {
      return(n)
    } else {
      paste(format(n, width = name_width), ':')
    }
  }
  format_value <- function(v) {
    if (is.na(v)) {
      return(NULL)
    } else if (is.null(v)) {
      return("NULL")
    }
    if (nchar(v) > 30L) {
      v <- paste0(strtrim(v, 26), "...")
    }
    v
  }

  json_s <- strsplit(json, split = ":", fixed = TRUE)
  name_width <- max(nchar(lapply(json_s, function(x) x[1])))

  ans <- lapply(json_s, function(x) {
    paste0(format_name(x[1]), format_value(x[2]))
  })

  paste(ans, collapse = "\n")
}
```
More functional and, I think, easier to modify in the future.

Reply:
I havent had a chance to see this proposition. Just to let you know. Will have a look tomorrow after the daily

Reply:
thanks for the input - I used this code and simplified it (only for centering) in this commit a85455764676ada29870b01ee7c059bd564e6cd8